### PR TITLE
[soltest] Improve interactive update routine

### PR DESCRIPTION
**Pull request by erak**
### Description

This PR removes the need to always match the size of expected and obtained results of semantic test function calls. The following test throws because they outputs don't match:
```
contract C {
    function f() public returns(uint d) {
        return true ? 5 : 10;
    }
}
// ====
// EVMVersion: >byzantium
// ----
// f() -> 
```
Exception:

`Encoding does not match byte range: the call returned 32 bytes, but 0 bytes were expected.`

### Solution

Since we can make use of the end-to-end execution framework, it's possible to get type information for the returned bytes from the contract ABI. These type are then used to format the output and make updating the test possible.

With these changes it will be possible to still auto-update the test without throwing an exception. A warning will be reported instead such the user can react accordingly:
```
Contract:
  contract C {
      function f() public returns(uint d) {
          return true ? 5 : 10;
      }
  }
  // ====
  // EVMVersion: >byzantium

  Expected result:
  // f() -> 

  Obtained result:
  // f() -> 5
  Warning: Encoding does not match byte range. The call returned 32 bytes, but 0 bytes were expected.
  
  Attention: Updates on the test will apply the detected format displayed.

```
### Output

isoltest

### CODE REVIEW

File: test/libsolidity/SemanticTest.cpp [hunk 1]
```diff
@@ -80,12 +80,19 @@ bool SemanticTest::run(ostream& _stream, string const& _linePrefix, bool _format
 	{
 		AnsiColorized(_stream, _formatted, {BOLD, CYAN}) << _linePrefix << "Expected result:" << endl;
 		for (auto const& test: m_tests)
-			_stream << test.format(_linePrefix, false, _formatted) << endl;
+		{
+			ErrorReporter errorReporter;
+			_stream << test.format(errorReporter, _linePrefix, false, _formatted) << endl;
+			errorReporter.print(_stream, _formatted, _linePrefix);
```
Comment (erak) [inline test/libsolidity/SemanticTest.cpp]:
```suggestion
			_stream << errorReporter.format(_linePrefix, _formatted);
```
This line will look like that in the end.

File: test/libsolidity/util/TestFunctionCall.cpp [hunk 2]
```diff
@@ -92,47 +98,49 @@ string TestFunctionCall::format(string const& _linePrefix, bool const _renderRes
...
 			}
 		}
 	};
 
-	if (m_call.displayMode == FunctionCall::DisplayMode::SingleLine)
-		formatOutput(true);
-	else
-		formatOutput(false);
-	return _stream.str();
+	formatOutput(m_call.displayMode == FunctionCall::DisplayMode::SingleLine);
```
Comment (erak) [inline test/libsolidity/util/TestFunctionCall.cpp:171]: Unrelated change.

File: test/libsolidity/util/TestFunctionCall.cpp [hunk 3]
```diff
@@ -92,113 +137,210 @@ string TestFunctionCall::format(string const& _linePrefix, bool const _renderRes
...
+		break;
+	case ABIType::None:
+		break;
 	}
-	return resultStream.str();
+	return os.str();
 }
 
-string TestFunctionCall::formatRawParameters(dev::solidity::test::ParameterList const& _params, std::string const& _linePrefix) const
+string TestFunctionCall::formatRawParameters(
```
Comment (erak) [inline test/libsolidity/util/TestFunctionCall.cpp:330]: Unrelated refactoring.

File: test/libsolidity/util/TestFunctionCall.h [hunk 4]
```diff
@@ -34,6 +35,99 @@ namespace solidity
...
+
+	/// Prints all errors depending on their type using ANSI colorized output.
+	/// It will be used to print notices, warnings and errors during the
+	/// interactive update process.
+	std::string format(std::string const& _linePrefix, bool _formatted)
+	{
+		std::stringstream os;
+		for (auto const& error: m_errors)
+		{
+			switch (error.type)
```
Comment (leonardoalt) [inline test/libsolidity/util/TestFunctionCall.h:101]: Could also just use two local variables for color and type string, and have only one `AnsiColorized` after the switch, but also fine with the current version.

File: test/libsolidity/SemanticTest.cpp [hunk 5]
```diff
@@ -74,6 +74,7 @@ bool SemanticTest::run(ostream& _stream, string const& _linePrefix, bool _format
 
 		test.setFailure(!m_transactionSuccessful);
 		test.setRawBytes(std::move(output));
+		test.setContractABI(m_compiler.contractABI(m_compiler.lastContractName()));
```
Comment (leonardoalt) [inline test/libsolidity/SemanticTest.cpp:77]: What about tests with multiple contracts?

File: test/libsolidity/util/TestFunctionCall.cpp [hunk 6]
```diff
@@ -22,12 +27,51 @@ using namespace solidity;
 using namespace dev::solidity::test;
 using namespace std;
 
-string TestFunctionCall::format(string const& _linePrefix, bool const _renderResult, bool const _highlight) const
+namespace
+{
+	/// Translates Solidity's ABI types into the internal type representation of
+	/// soltest.
+	auto getContractABITypes(string const& _type) -> vector<ABIType>
```
Comment (leonardoalt) [inline test/libsolidity/util/TestFunctionCall.cpp]:
```suggestion
	auto contractABITypes(string const& _type) -> vector<ABIType>
```

File: test/libsolidity/util/TestFunctionCall.cpp [hunk 7]
```diff
@@ -22,12 +27,51 @@ using namespace solidity;
...
+	{
+		regex boolType{"(bool)"};
+		regex unsignedType{"(uint\\d*)"};
```
Comment (leonardoalt) [inline test/libsolidity/util/TestFunctionCall.cpp]:
```suggestion
		regex uintType{"(uint\\d*)"};
```

File: test/libsolidity/util/TestFunctionCall.cpp [hunk 8]
```diff
@@ -22,12 +27,51 @@ using namespace solidity;
...
+		regex signedType{"(int\\d*)"};
```
Comment (leonardoalt) [inline test/libsolidity/util/TestFunctionCall.cpp]:
```suggestion
		regex intType{"(int\\d*)"};
```

File: test/libsolidity/util/TestFunctionCall.cpp [hunk 9]
```diff
@@ -125,88 +174,173 @@ string TestFunctionCall::format(
...
+	stringstream os;
+	string functionName{_signature.substr(0, _signature.find("("))};
 
-	if (_bytes.empty())
-		return {};
 	auto sizeFold = [](size_t const _a, Parameter const& _b) { return _a + _b.abiType.size; };
 	size_t encodingSize = std::accumulate(_params.begin(), _params.end(), size_t{0}, sizeFold);
 
-	if (encodingSize != _bytes.size())
+	/// Infere type from Contract ABI. Used to generate values for
```
Comment (leonardoalt) [inline test/libsolidity/util/TestFunctionCall.cpp]: Infere -> Infer

File: test/libsolidity/SemanticTest.cpp (hunk 5 above)
Comment (erak) [reply test/libsolidity/SemanticTest.cpp:77]: Right now, you can't call functions on a contract that is not the recently compiled one directly, which means that only its contract ABI is of interest.

File: test/libsolidity/util/TestFunctionCall.h (hunk 4 above)
Comment (erak) [reply test/libsolidity/util/TestFunctionCall.h:101]: Mmh, then we would need a temporary `const char*` and I'm not sure if a few lines less can justify additional state. I think, I would stick to the current solution.

File: test/libsolidity/util/TestFunctionCall.cpp [hunk 10]
```diff
@@ -22,6 +27,39 @@ using namespace solidity;
 using namespace dev::solidity::test;
 using namespace std;
 
+namespace
+{
+	/// Translates Solidity's ABI types into the internal type representation of
```
Comment (chriseth) [inline test/libsolidity/util/TestFunctionCall.cpp]: We usually do not indent inside namespaces.

File: test/libsolidity/util/TestFunctionCall.cpp [hunk 11]
```diff
@@ -22,6 +27,39 @@ using namespace solidity;
...
+	/// soltest.
+	auto contractABITypes(string const& _type) -> vector<ABIType>
+	{
+		regex boolType{"(bool)"};
```
Comment (chriseth) [inline test/libsolidity/util/TestFunctionCall.cpp]: static?

File: test/libsolidity/util/TestFunctionCall.cpp [hunk 12]
```diff
@@ -125,88 +174,173 @@ string TestFunctionCall::format(
...
+			}
+
+
+	/// If parameter count does not match, take types defined by ABI, but only
+	/// if the contract ABI is defined (needed for format tests where the actual
+	/// result does not matter).
+	ParameterList preferredParams;
+	if (m_contractABI && (_params.size() != abiParams.size()))
+	{
 		_errorReporter.warning(
```
Comment (chriseth) [inline test/libsolidity/util/TestFunctionCall.cpp:210]: Is this error really that important? Won't we display an error already due to the values not matching? Perhaps an assertion can be put here that checks that the sizes are the same if the values are the same.

File: test/libsolidity/util/TestFunctionCall.cpp (hunk 10 above)
Comment (erak) [reply test/libsolidity/util/TestFunctionCall.cpp]: Thanks.

File: test/libsolidity/util/TestFunctionCall.cpp (hunk 11 above)
Comment (erak) [reply test/libsolidity/util/TestFunctionCall.cpp]: Yes, that makes sense.

File: test/libsolidity/util/TestFunctionCall.cpp (hunk 12 above)
Comment (erak) [reply test/libsolidity/util/TestFunctionCall.cpp:210]: Yes, it would result in an error already. But it turned out, that it makes sense to have this information (expected vs. actual size) when developing more complex tests (e.g. calldata arrays). The idea behind the slightly over-engineered error reporter, which is also part of this PR, was to give the user as many information as needed. I'm thinking about test-driven development, which would also benefit from a verbose output. 

One drawback I can see here, is the mapping from a Solidity type to the test-internal ABIType representation, which is kind of incomplete right now (reference types need to be fixed). But still, you would have the warning even if the auto-corrected value or its size is incorrect.

Comment (chriseth) [reply test/libsolidity/util/TestFunctionCall.cpp:210]: Ok!

**Review by chriseth — APPROVED** [review]
